clippy_lints/src/needless_borrow.rs

```diff
@@ -86,50 +108,69 @@ impl<'tcx> LateLintPass<'tcx> for NeedlessBorrow {
         }
     }
     fn check_pat(&mut self, cx: &LateContext<'tcx>, pat: &'tcx Pat<'_>) {
-        if pat.span.from_expansion() || self.derived_item.is_some() {
-            return;
-        }
         if_chain! {
-            if let PatKind::Binding(BindingAnnotation::Ref, .., name, _) = pat.kind;
-            if let ty::Ref(_, tam, mutbl) = *cx.typeck_results().pat_ty(pat).kind();
-            if mutbl == Mutability::Not;
-            if let ty::Ref(_, _, mutbl) = *tam.kind();
+            if let PatKind::Binding(BindingAnnotation::Ref, id, name, _) = pat.kind;
+            if !in_external_macro(cx.sess(), pat.span);
+            if let ty::Ref(_, tam, _) = *cx.typeck_results().pat_ty(pat).kind();
             // only lint immutable refs, because borrowed `&mut T` cannot be moved out
-            if mutbl == Mutability::Not;
+            if let ty::Ref(_, _, Mutability::Not) = *tam.kind();
             then {
+                let (span, search_loc) = match get_pat_usable_location(cx.tcx, pat) {
+                    Some(x) => x,
+                    None => return,
+                };
+                let ctxt = span.ctxt();
```

Comment:
Wouldn't `pat.span.ctxt()` be more appropriate? See also comment on one of the tests.

Reply:
This is in reference to `ctxt` usages other than the one immediately below :)

Reply:
Both `span.ctxt()` and `pat.span.ctxt()` are compared for equality.